clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2Credentials.java

```diff
@@ -679,4 +573,130 @@ public void patch(
       }
     }
   }
+
+  /**
+   * Handles validating whether kubernetes kinds should be cached by the current account, and
+   * whether metrics are enabled for the account.
+   */
+  private class KindValidator {
+    private final Supplier<String> checkNamespace = Suppliers.memoize(this::computeCheckNamespace);
+    private final Map<KubernetesKind, Boolean> readableKinds = new ConcurrentHashMap<>();
+    private final Supplier<Boolean> metricsReadable = Suppliers.memoize(this::checkMetricsReadable);
+
+    private final Set<KubernetesKind> kinds;
+    private final Set<KubernetesKind> omitKinds;
+
+    KindValidator(Set<KubernetesKind> kinds, Set<KubernetesKind> omitKinds) {
+      this.kinds = kinds;
+      this.omitKinds = omitKinds;
+    }
+
+    private String getCheckNamespace() {
+      return checkNamespace.get();
+    }
+
+    private String computeCheckNamespace() {
+      List<String> namespaces = getDeclaredNamespaces();
+
+      if (namespaces.isEmpty()) {
+        log.warn(
+            "There are no namespaces configured (or loadable) -- please check that the list of 'omitNamespaces' for account '"
+                + accountName
+                + "' doesn't prevent access from all namespaces in this cluster, or that the cluster is reachable.");
+        return null;
+      }
+
+      // we are making the assumption that the roles granted to spinnaker for this account in all
+      // namespaces are identical.
+      // otherwise, checking all namespaces for all kinds is too expensive in large clusters
+      // (imagine
+      // a cluster with 100s of namespaces).
+      return namespaces.get(0);
+    }
+
+    private boolean skipPermissionChecks() {
+      // checkPermissionsOnStartup exists from when permission checks were done at startup (and took
+      // a long time); this flag was added to skip the checks and assume all kinds were readable.
+      // Now that permissions are checked on-the-fly, this flag is probably not necessary, but for
```

Comment:
> but for now...

If possible, I like to put clear removal timelines in the comment like "this can be removed for the 1.18 release unless we see major problems with the new startup behavior in 1.17". So that when we forget about this and stumble across it again in 2021 we can feel confident in removing it. But maybe this requires more discussion with customers so a hard timeline isn't appropriate.

Reply:
Added some more detail